iohk/NixOps.hs

```diff
@@ -403,156 +626,374 @@ exists o c@NixopsConfig{..} = do
+
+stop :: Options -> NixopsConfig -> IO ()
+stop o c = echo "Stopping nodes..."
+  >> parallelSSH o c "systemctl" ["stop", "cardano-node"]
+
+defLogs, profLogs :: [(Text, Text -> Text)]
+defLogs =
+    [ ("/var/lib/cardano-node/node.log", (<> ".log"))
+    , ("/var/lib/cardano-node/jsonLog.json", (<> ".json"))
+    , ("/var/lib/cardano-node/time-slave.log", (<> "-ts.log"))
+    , ("/var/log/saALL", (<> ".sar"))
+    ]
+profLogs =
+    [ ("/var/lib/cardano-node/cardano-node.prof", (<> ".prof"))
+    , ("/var/lib/cardano-node/cardano-node.hp", (<> ".hp"))
+    -- in fact, if there's a heap profile then there's no eventlog and vice versa
+    -- but scp will just say "not found" and it's all good
+    , ("/var/lib/cardano-node/cardano-node.eventlog", (<> ".eventlog"))
+    ]
+
+start :: Options -> NixopsConfig -> IO ()
+start o c =
+  parallelSSH o c "bash" ["-c", Arg $ "'" <> rmCmd <> "; " <> startCmd <> "'"]
+  where
+    rmCmd = foldl (\str (f, _) -> str <> " " <> f) "rm -f" logs
+    startCmd = "systemctl start cardano-node"
+    logs = mconcat [ defLogs, profLogs ]
+
+date :: Options -> NixopsConfig -> IO ()
+date o c = parallelIO o c $
+  \n -> ssh' o c "date" [] n
+  (\out -> TIO.putStrLn $ fromNodeName n <> ": " <> out)
+
+wipeJournals :: Options -> NixopsConfig -> IO ()
+wipeJournals o c@NixopsConfig{..} = do
+  echo "Wiping journals on cluster.."
+  parallelSSH o c "bash"
+    ["-c", "'systemctl --quiet stop systemd-journald && rm -f /var/log/journal/*/* && systemctl start systemd-journald && sleep 1 && systemctl restart nix-daemon'"]
+  echo "Done."
+
+getJournals :: Options -> NixopsConfig -> IO ()
+getJournals o c@NixopsConfig{..} = do
+  let nodes = nodeNames o c
+
+  echo "Dumping journald logs on cluster.."
+  parallelSSH o c "bash"
+    ["-c", "'rm -f log && journalctl -u cardano-node > log'"]
+
+  echo "Obtaining dumped journals.."
+  let outfiles  = format ("log-cardano-node-"%s%".journal") . fromNodeName <$> nodes
+  parallelIO' o c (flip zip outfiles) $
+    \(node, outfile) -> scpFromNode o c node "log" outfile
+  timeStr <- T.pack . timePrint ISO8601_DateAndTime <$> dateCurrent
+
+  let archive   = format ("journals-"%s%"-"%s%"-"%s%".tgz") (lowerShowT cEnvironment) (fromNixopsDepl cName) timeStr
+  printf ("Packing journals into "%s%"\n") archive
+  cmd o "tar" (["czf", archive, "--force-local"] <> outfiles)
+  cmd o "rm" $ "-f" : outfiles
+  echo "Done."
+
+confirmOrTerminate :: Text -> IO ()
+confirmOrTerminate question = do
+  echo $ unsafeTextToLine question <> "  Enter 'yes' to proceed:"
+  reply <- readline
+  unless (reply == Just "yes") $ do
+    echo "User declined to proceed, exiting."
+    exit $ ExitFailure 1
+
+wipeNodeDBs :: Options -> NixopsConfig -> IO ()
+wipeNodeDBs o c@NixopsConfig{..} = do
+  confirmOrTerminate "Wipe node DBs on the entire cluster?"
+  parallelSSH o c "rm" ["-rf", "/var/lib/cardano-node"]
+  echo "Done."
+
+updateNixops :: Options -> NixopsConfig -> IO ()
```

Comment:
Instead of this, we should just use `nix-build -A nixops` and modify what's there to be used for deployments

Reply:
I once tried to override `nixops`, but it doesn't seem like a normal derivation..  I guess I need someone's (@cleverca22 ?) help here.

Reply:
one thought i had is to just use IFD, exactly the same way we do for stack2nix

Reply:
Right, that's how it ended up being implemented.